Unify M600 and M125 pause features

tcm0116 opened the pull request:
@Roxy-3D requested the ability to prime the nozzle when resuming a print. While looking into this request, it occurred to me that the recently added M125 code was was almost copy-and-pased from M600. As such, I decided it would be helpful to merge them together into a single implementation in order to have the flexibility to provide the benefits of both without a bunch of code duplication.

- Renamed the FILAMENT_CHANGE_FEATURE to ADVANCED_PAUSE_FEATURE.
- Added U parameter to M600 to allow for specifying the unload length separately from the load length.
- Added the B parameter to M600 to allow for specifying how many times to beep when the reload is ready.
- Disabled beeping for M600 when called from the LCD.
- Moved the heater timeout controls to ThermalManager to provide for a more generic implementation.
- Modified the LCD to flash the target temperature if the heater has been put into an idle state.
- As part of the unification, M24 will now wait for the heaters to re-heat and present the user with the option for priming the nozzle (if equipped with an LCD).

I tried to give this a fair amount of testing, but I only have one printer with a 20x4 LCD. As such, it would be helpful if this could get some more testing on different configurations, especially the target temperature idle flashing on the DOGM display when the heaters have become idle after pausing.

tcm0116 committed:
Unify M600 and M125 pause features

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5550,38 +5550,279 @@ inline void gcode_M17() {
+    }
+  }
+
+  static bool pause_print(float retract, float z_lift, float x_pos, float y_pos,
```
For `float` arguments on AVR, `const float &` is preferred whenever possible.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -8566,231 +8753,77 @@ inline void gcode_M503() {
-      #if FILAMENT_CHANGE_LOAD_LENGTH > 0
-        + FILAMENT_CHANGE_LOAD_LENGTH
+    const float load_length = code_seen('L') ? code_value_axis_units(E_AXIS) : 0
+      #if defined(FILAMENT_CHANGE_LOAD_LENGTH)
```
`#ifdef` should be used instead of `#if defined(...)` for cases like this.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -8566,231 +8753,77 @@ inline void gcode_M503() {
-    #endif
-    stepper.synchronize();
+    const int beep_count = code_seen('B') ? code_value_int() :
+      #if defined(FILAMENT_CHANGE_NUMBER_OF_ALERT_BEEPS)
```
`#ifdef`…

thinkyhead commented:
This would be a good opportunity also to unify this with `NOZZLE_PARK_FEATURE`.

tcm0116 commented:
> This would be a good opportunity also to unify this with NOZZLE_PARK_FEATURE

I could see how it could make use of that feature for moving to the park location. I'll look into the idea.

tcm0116 commented:
@thinkyhead Can you provide any guidance on what you're thinking with the NOZZLE_PARK_FEATURE?

tcm0116 committed:
Cleanup per thinkyhead's comments

thinkyhead commented:
We can get into it more after the release.

tcm0116 commented:
@thinkyhead I understand if you want to push this off to the next release. However, I think there are some fixes in this pull request that would be good to include in this release. For example, this pull request fixes an issue with the Filament Reload screen that was modified to show the extruder temperature, but didn't actually update. Also, I fixed several issues related to cancelling a SD print, such as not shutting off the heaters or fans.

Would you like for me to extract some of the smaller bug fixes and submit them as separate pull requests?

Roxy-3D commented:
> I understand if you want to push this off to the next release. However, I think there are some fixes in this pull request that would be good to include in this release. 

@tcm0116  I've been super busy looking for an elusive bug.   But I think I have that one figured out and fixed.  What can I do to help?

tcm0116 commented:
@Roxy-3D not sure. If @thinkyhead wants me to extract the few small bug fixes from this pull request, I can get that taken care of no problem. 

Maybe it's about time to start having an architecture discussion. I'm thinking that @thinkyhead is wanting to move some of these types of features into the Nozzle class, which makes a lot of sense. However, it might be helpful if we put together a diagram of sorts to help us determine which features and capabilities will go into which classes. With that as an end point, we can start refactoring the code one piece at a time into a fully object-oriented implementation. 

Furthermore, I'm stewing on an idea to support different tool heads without having to re-load the firmware. If you consider LulzBot's TAZ-6 design with their various tool heads, it could be a compelling feature. In my case, I'm going with the E3D Chimera & Cyclops Legendary Pack, and I just like the idea of not having to re-flash the firmware when I switch between the two. The Nozzle class seems like a good place to handle this, but it's going to make Configuration.h rather interesting.

Roxy-3D commented:
> Maybe it's about time to start having an architecture discussion. I'm thinking that @thinkyhead is wanting to move some of these types of features into the Nozzle class, which makes a lot of sense. However, it might be helpful if we put together a diagram of sorts to help us determine which features and capabilities will go into which classes. With that as an end point, we can start refactoring the code one piece at a time into a fully object-oriented implementation.

All that is fine and good...  But if we can do a couple small tweaks and make this suitable to be included in the 'Golden Master' I think we should do it.   And we can have an architectural discussion about how to organize things better for v1.1.1 after this Release Candidate gets promoted.

tcm0116 commented:
@thinkyhead Are we at a point where we can resume the discussion about this? With the PR for M701 and M702 (#6493) making yet another version of the filament change code, I think we're starting to see a strong case for unifying all of these very related features in order to cut down on the amount of duplicate code floating around. There were also some heater pausing features added in #6622 for BLTouch that overlap some of the work in this PR related to turning off the heaters during an extended pause.

I still don't exactly follow your comment about unifying this with the NOZZLE_PARK_FEATURE, so I'd like to have some discussion about what you're thinking with that.

Roxy-3D commented:
@tcm0116    Are there any pieces of code that can be combined to shrink redundancy?   That is not an architectural discussion...  But it probably will kick of some discussion in that direction.

tcm0116 commented:
@Roxy-3D when M125 was added, the code for M600 was duplicated and massaged a bit. The aim of this PR was to consolidate them back into a single back-end that was shared between the M125 and M600 commands. I believe it wouldn't be much more effort to make use of the same shared functionality for the M701 and M702 filament load and unload commands from #6493.

tcm0116 committed:
Merge branch 'advanced_pause2' into bugfix-1.1.x

tcm0116 committed:
Rename filament_change_menu_response to advanced_pause_menu_response

tcm0116 commented:
I've updated this PR to advance it to the current bugfix-1.1.x branch. As part of that, I've merged the heater pause capability that @bgort added in #6622 with the heater idle timeout capability I had originally added. I'd still like to merge in M701 and M702 from #6493, but I'd like to get some more feedback on what I've done so far before I proceed.

I still need someone to test the target temperature flashing on the DOGM display when the heaters and/or bed are paused. That can be done by starting a print from SD and then pausing it via the LCD. After the configured timeout, the target temperatures for the extruders should start flashing and the actual temperature will start dropping. Testing the bed temperature flashing should be able to be accomplished by enabling the pause while probing feature and then it should flash during each probe event.

bgort commented:
Looking good....

bgort commented on Marlin/Marlin_main.cpp:
```diff
@@ -2834,7 +2834,7 @@ static void do_homing_move(const AxisEnum axis, float distance, float fr_mm_s=0.
   stepper.synchronize();
 
-  #if QUIET_PROBING
+  #if HAS_BED_PROBEQUIET_PROBING
```
Typo?

tcm0116 replied:
Oops. Yes.

bgort replied:
Hmm.  Wonder if it's better to just add a conditional so that QUIET_PROBING can't be set if HAS_BED_PROBE isn't?

tcm0116 replied:
I was actually thinking the same thing, so I'll look into it.

Roxy-3D commented:
How do we get this Pull Request woken up and moving forward???   This is good stuff.  And other users are asking for it too!  https://github.com/MarlinFirmware/Marlin/issues/6763

fiveangle commented:
I probably shouildn't do this... https://github.com/MarlinFirmware/Marlin/issues/5982#issuecomment-302154942

Feature creep is baaaaaaaaaad, mkaaaaaaay ? ;)

tcm0116 committed:
Include HAS_BED_PROBE in QUIET_PROBING

tcm0116 committed:
Merge remote-tracking branch 'upstream/bugfix-1.1.x' into HEAD

tcm0116 commented:
@thinkyhead Would it be possible to remove the "Don't Merge" status label from this pull request? There is obviously interest in moving forward with it, and you had even said that we would get it to it more after the release.

bgort commented:
Looks like it was rebased and is conflict free; just removed the 'needs: rebase' label.

I'll review today, too.

Roxy-3D commented:
This affects enough files...   Besides reviewing it, I'm thinking we should load it up and test it.   I certainly know how M600 should behave.   But some clues on what to look for on M125 would be helpful.

bgort commented:
Yeah - was thinking the same thing.  Should definitely be tested before merging.

bgort commented:
Now looks like there's a conflict.  Wonder if Github is playing tricks on me... or maybe something was just merged and created a conflict.

Roxy-3D commented:
I wonder if we can come up with another name instead of  ADVANCED_PAUSE_FEATURE.   The problem is what comes next?   The super advanced pause feature ?    We are not going to have both a Pause Feature and an Advanced Pause feature, right?    Can we just call this the Park and Pause Feature ?

tcm0116 commented:
> or maybe something was just merged and created a conflict

Yep. #6773. I can merge again later this evening. 

>Can we just call this the Park and Pause Feature ?

Could that (and advanced pause, for that matter) lead to confusion since this has the filament change incorporated into it?

Roxy-3D commented:
I think the 'Right' way to handle that is if either feature is turned on...    What ever is needed gets pulled in.   But it kind of seems like Park & Pause is a subset of Filament Change.   So maybe the right thing to do is require (or automatically turn on) Park & Pause if Filament Change is enabled.

Your thoughts???

Roxy-3D commented:
@tcm0116  @bgort    I just tried to bring this up and test it....   I was going to re-add the 'Needs Rebase' label but I see you beat me to it.   I'm having trouble compiling because the Planner::abl_enabled not being part of the class.

Rather than spend the time to fight this...   I'll bring it up first thing tomorrow morning if we can get it rebased.    Also...   There are some new example files that were not around when this branch was made.  Specifically FolgerTech-i3-2020 and gCreate_gMax1.5+ that will need the configuration files brought into compilance.   

My suggestion is let's test it (as quickly as practical after it is rebased) to make sure things pretty much work as expected and don't break anything.   And then we merge it before we start getting conflicts.  We can fine tune the conditional compilation and names (if needed) later.   Right now...  It looks to me like FILAMENT_CHANGE_FEATURE should require the PARK_AND_PAUSE_FEATURE.   (If I can be allowed to twist a everybody's arm a little...  I think PARK_AND_PAUSE_FEATURE as a name is very easy for people to understand.  

One thing I did notice is it might make sense to move the NOZZLE_PARK_FEATURE over to Configuration_adv.h and make its parameters part of the PARK_AND_PAUSE_FEATURE.   If we don't do that, we are going to be causing some confusion...

bgort commented:
Makes good sense to me.  I can test it once the conflict is resolved, too.

tcm0116 committed:
Merge remote-tracking branch 'upstream/bugfix-1.1.x' into advanced_pause2

tcm0116 commented:
I've re-based it again, so it should be ready to test.

If we want to re-name it, I'm fine with that, but let's get some testing on it and I'll incorporate the name change with any issues that might be found.

Roxy-3D commented:
@tcm0116   May I have write privileges to this repository?  I'll get things right and test it...  And if you approve the changes you can do a Pull Request and I'll merge it...   Does that work for you?  (I already changed the #define names in the gMax example files...  It's easier just to make the changes and bring everything up than to try to get consensus on each step...)

Right now I've got an is_idle() not in the Planner:: scope problem...  That will be the next thing that gets some correction...

tcm0116 commented:
@Roxy-3D  I added you as a collaborator to my fork. Let me know if that works.

Roxy-3D commented:
I'm almost positive that works...  But I won't start making changes until
the morning...   I've been playing darts (on a Friday night).     Some
times we bet beers...

On Fri, May 19, 2017 at 10:44 PM, Thomas Moore <notifications@github.com>
wrote:

> @Roxy-3D <https://github.com/roxy-3d> I added you as a collaborator to my
> fork. Let me know if that works.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/6407#issuecomment-302848280>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIA9YhiI6Sf2zlHWR2z0dQRn13qZMiqJks5r7mG2gaJpZM4ND2fb>
> .
>

Roxy-3D committed:
Update gMax example file

Roxy-3D committed:
is_idle() is out of scope without the braces

Roxy-3D committed:
Convert FT-i3-2020 to Advance Pause names...

Roxy-3D commented:
@tcm0116   I have the example configuration files updated and consistent...   (I needed that to test.)   Those are saved so when we do the Pull Request, that doesn't have to be re-done. 

But M600 doesn't do anything.   I added some debug code so you can see what I'm talking about...  I homed the machine, I heated up both the bed and nozzle...   And the M600 just screams through all the steps, but the filament never retracts.   And it doesn't wait for the user.

M125 doesn't seem to do anything either.    

Is it possible the code is assuming a print is in progress?  Because M600 should work either way.

tcm0116 commented:
@Roxy-3D Line 5833 of Marlin_main is:

`if (!move_away_flag) return false; // nothing to pause`

This was in the original park on pause code, but I'm not sure about the original M600 code. In my testing, I actually started prints before testing the commands.

Roxy-3D commented:
> This was in the original park on pause code, but I'm not sure about the original M600 code. In my testing, I actually started prints before testing the commands.

:)   Yeah!  That is why it is important to have several people test things!   Pretty much...  I use M600 in the middle of a print so I can change colors.   But some times, I use it to switch colors before a print.   So figured I would start small and see what happened.   And the filament did not retract...

But now my printer is damaged https://github.com/MarlinFirmware/Marlin/issues/6807 so it will be a few days until I can recover.

Roxy-3D commented:
@tcm0116   Do you know how to fix this for the 'Not Printing' case?   Is that what you are working on?

tcm0116 committed:
Allow pause even if not printing

tcm0116 committed:
Merge remote-tracking branch 'upstream/bugfix-1.1.x' into HEAD

tcm0116 commented:
I've re-based this again and also incorporated the change requested by @Roxy-3D to allow filament change when printing is not in progress.

@Roxy-3D I reverted your debugging commits just to clean this up a bit. Hopefully we can get this tested, squashed, and merged without having to re-base yet again.

Roxy-3D commented:
Reverting the debug stuff makes good sense!
I'll load it in the morning and see how it behaves.  

Update:   I've been fighting my equipment all morning to get this loaded so I can test it.   But I'm out of time and have to take care of a few things.   With any luck, things will go smoother this afternoon when I get back and I'll be able to spend some time working with it.   If things go reasonable, I'll merge it then.

Roxy-3D commented:
Everything looks good and behaves as expected...   With one exception.  And the one exception isn't caused by this Pull Request.   

The one thing I noticed is the LCD Panel on M600 when it is allowed to timeout.    It displays the nozzle's temperature in the format Temperature E1:  194/210   

If the nozzle was not turned off...  The target temperature would be 210 (and will be once the user clicks the encoder wheel to continue).   But having the first number fall is confusing when the target is claimed to be much higher.   I'm guessing we would have to add a third number to make things make sense.   And I don't know that we want to do that....

THANKS @tcm0116  !!!!

UPDATE:   I've been playing some more with this...  Wouldn't it make sense to have the LCD Panel 'Pause Print' up higher in the menu system?   If the printer is printing...   It would be nice to pause even if the print isn't happening from an SD-Memory card.  (Right now...  The Pause feature is buried in the SD-Card menu.)

For right now...  This works fine for me.  I can just put big prints that I might want to pause at night time on an SD-Memory card.   But...  eventually...  It would be nice to be able to pause regardless how the printer is getting the GCode.

tcm0116 commented:
> The one thing I noticed is the LCD Panel on M600 when it is allowed to timeout. It displays the nozzle's temperature in the format Temperature E1: 194/210

The second number after the / is supposed to flash when the heater is timed out. This should happen when the temperature status is displayed in the LCD during a filament change or when resuming from a pause. The target temperature should also flash on the main screen if a SD card print is paused and the heater times out.

If the target temperature is not flashing, then there's an issue with this PR. Do you have a DOGLCD or ULTRA_LCD display?

Roxy-3D commented:
> If the target temperature is not flashing, then there's an issue with this PR. Do you have a DOGLCD or ULTRA_LCD display?

I'm not sure I noticed that.   I did notice the flashing number on the main status screen.   I have a Graphical LCD on the machine I did the testing on.    

Oh!   I just realized something....   The numbers I was talking about aren't part of the status screen.  They are special display screens done by the Filament Change code.    So...  we can look at those and see how bad it would be to duplicate the blinking there.  But for now...  The code does work!

PLEASE don't take my comments as critical of the work!!!!  This gives me what I need to pause long prints at night time!!!   But as always...  no matter how good the code is...  somebody will say "Yeah but it would be nicer if...."

tcm0116 commented:
@Roxy-3D I **am** taking your comments as critical of my work, but from the perspective that I want my work to function correctly. I greatly appreciate you helping me with this, so keep beating on it 😁

I'm surprised that the temps flash on the main screen but not on the pause screens since both places call the same function to draw the temperatures. When the heaters time out while the pause screen is displayed during a filament change, does the display update automatically such that the temperature value changes without you having to interact with the printer in any way, or do you have to press the encoder button to see the temperature change?

Roxy-3D commented:
> I'm surprised that the temps flash on the main screen but not on the pause screens since both places call the same function to draw the temperatures. 

It is very possible I'm wrong about this...   I'll have a chance to re-test this later today and I'll keep an eye on that number to see what it does.   But in M600, when the nozzle times out and turns off.   You see the actual temperature number (the first one) slowly start dropping.  If I remember right...  The second number (the target teemperature) just stays at the original number.   If it was blinking...  I missed that...

But I'll be able to provide more detail later.

floyd871 commented:
With this Merge you have overwritten two entries in many language files with the english one!!
Please revert that.

#define MSG_FILAMENT_CHANGE_HEADER          _UxGT("PRINT PAUSED")
#define MSG_FILAMENT_CHANGE_OPTION_HEADER _UxGT("RESUME OPTIONS:")

tcm0116 commented:
@floyd871 that was by design. The original text was specific to the M600 filament change, but don't make sense to the M125 pause since it now uses some of those screens when the print is resumed. It's my understanding that there are people that periodically update the translations as necessary. If you want to provide a translation for your language, I'm sure we could get it incorporated.

floyd871 commented:
I see, that makes sense, but then it would be better if you have renamed the label as well.
"Filament change" doesn't match "print paused".... just my opinion.

tcm0116 commented:
I thought the same thing when you posted them. I tried to change
everything, but I guess I missed those.

Roxy-3D commented:
Well...  Now that the bulk of the changes are merged....  It is easy to followup with a couple of small patches.   Somebody is helping!  https://github.com/MarlinFirmware/Marlin/pull/6871

Roxy-3D commented:
@tcm0116   I've been using the Advanced Pause and new Filament Change for 2 days now.  I have yet to get the print I wanted done.  (It is a very difficult print)   But the new Pause code seems to do exactly what it is supposed to do.  I have paused it several times and changed the filament several times.

Unfortunately...  I thought the last print was going to succeed, but I paused it for a couple of hours that I had to be away from the printer.   When I cane back, the ABS plastic I  was using broke away from the glass and lost adhesion.    Soooo...   Restarting it with PLA plastic on my FrankenStein printer....

tcm0116 commented:
@Roxy-3D That's great to hear! Did you happen to remember to check for the
flashing target temperature on the filament change LCD screens?

Roxy-3D commented:
No!   I forgot...  But I'll try to remember to do that next time I pause something...

Roxy-3D commented:
@tcm0116 

As it turns out...   We still have a little bit of work to do!   :)

When giving the printer a M600...    The second number doesn't blink.

tcm0116 commented:
@Roxy-3D does the first number update as the temperature is dropping without you having to do anything like move the encoder?

Roxy-3D commented:
> does the first number update as the temperature is dropping without you having to do anything like move the encoder?

Yes!   It is slow on my printer...  but after 30 or 40 seconds, it is clear that it is dropping....

tcm0116 commented:
@Roxy-3D I put together a change at https://github.com/tcm0116/Marlin/commit/f3528b9b42c2940b4ad0eda8701d814e04ad1d10 if you have a chance to give it a try.

Roxy-3D commented:
I can grab your pause_flash branch, right?   But...   should we edit this?  We have nested       #if ENABLED(ADVANCED_PAUSE_FEATURE)

```cpp
 #if ENABLED(ADVANCED_PAUSE_FEATURE)

    static void lcd_implementation_hotend_status(const uint8_t row) {
      row_y1 = row * row_height + 1;
      row_y2 = row_y1 + row_height - 1;

      if (!PAGE_CONTAINS(row_y1 + 1, row_y2 + 2)) return;

      u8g.setPrintPos(LCD_PIXEL_WIDTH - 11 * (DOG_CHAR_WIDTH), row_y2);
      lcd_print('E');
      lcd_print((char)('1' + active_extruder));
      lcd_print(' ');
      lcd_print(itostr3(thermalManager.degHotend(active_extruder)));
      lcd_print('/');
      
      #if ENABLED(ADVANCED_PAUSE_FEATURE)
        if (lcd_blink() || !thermalManager.is_heater_idle(active_extruder))
      #endif	
      lcd_print(itostr3(thermalManager.degTargetHotend(active_extruder)));
    }

  #endif // ADVANCED_PAUSE_FEATURE
```

tcm0116 commented:
@Roxy-3D Oops. Updated in the branch https://github.com/tcm0116/Marlin/tree/pause_flash

Roxy-3D commented:
Grabbing it now...

Roxy-3D commented:
It works on a Graphical Display!!!   I don't have a way to test a 20x4 right now.

tcm0116 commented:
I can test it on my printer.

thinkyhead commented:
Seems there is a problem. The filament change feature was broken by this PR, apparently. If the nozzle is cooling down when you start `M600`, but it is above the extrusion temperature, then `M600` is allowed to proceed. However, this PR removes the code from `M600` that was put in place to ensure that the temperature won't drop below safe extrusion temperature before first ejection and during loading.

Let's restore that code to `M600`, because otherwise the temperature just drops and never gets set to the extrusion temperature again.

```cpp
// Show "wait for heating"
lcd_filament_change_show_message(FILAMENT_CHANGE_MESSAGE_WAIT_FOR_NOZZLES_TO_HEAT);

wait_for_heatup = true;
while (wait_for_heatup) {
  idle();
  wait_for_heatup = false;
  HOTEND_LOOP() {
    if (abs(thermalManager.degHotend(e) - temps[e]) > 3) {
      wait_for_heatup = true;
      break;
    }
  }
}
```

thinkyhead commented:
Moreover, the original code I believe took into account that you can break out of a wait for temperature loop prematurely via `M108` — so we will need to make sure that after waiting for temperature that the code will abort if the target temperature has still not been reached.

I would advise combing the previous `gcode_M600` code and making sure that the current code reproduces beat-for-beat everything that the previous code did. It is not enough, for example, to put up the `wait_for_heating` screen. The `M600` code was responsible for also tracking the temperature and moving forward once the target was reached. The approach with `M600` was to make the gcode handler the controller of the LCD, in contrast to procedures which are initiated from the LCD.

Roxy-3D commented:
Maybe we should revert this?   And then move forward with that in mind?  The main purpose of the commit was to collapse duplicate code.

thinkyhead commented:
I think at this point it will be less work to plough forward.

tcm0116 commented:
> However, this PR removes the code from M600 that was put in place to ensure that the temperature won't drop below safe extrusion temperature before first ejection and during loading.

I don't believe this is a true statement. As can be seen in the snippet below from `resume_print` from the current bugfix-1.1.x branch, the code you've referenced is still present. The difference is that the old code stored the target temperatures prior to turning off the heaters, whereas the new code directly looks at `thermalManager.degTargetHotend(e)`. The only difference this could result in is if the user changes the target temperature while the print is paused for the M600, but I don't think that's possible since the LCD is blocked by the pause screens and the parser is blocked by the call to `gcode_M600`. The reason for this change was because I added the ability to put the heaters into an idle state directly in the Thermal Manager, so it was no longer necessary for M600 to keep track of the original target temperature.

> If the nozzle is cooling down when you start M600, but it is above the extrusion temperature, then M600 is allowed to proceed.

I'm 100% certain that the old M600 code would do the exact same thing.

> Moreover, the original code I believe took into account that you can break out of a wait for temperature loop prematurely via M108

The `while (wait_for_heatup)` line in `resume_print` maintains that capability in the exact same manner as the original M600 code. However, there appears to logic error in that `resume_print` (and the old `gcode_M600`) is spinning on `wait_for_heatup`, but then manipulates `wait_for_heatup` within the loop. If `resume_print` enters the while loop while `wait_for_heatup` is **true**, but then the LCD click or M108 resets `wait_for_heatup` to **false** during the call to `idle()` in the loop, then that transition to **false** will be lost because `resume_print` overwrites it blindly. 

```
#if ENABLED(ULTIPANEL)
  // Show "wait for heating"
  lcd_advanced_pause_show_message(ADVANCED_PAUSE_MESSAGE_WAIT_FOR_NOZZLES_TO_HEAT);
#endif

wait_for_heatup = true;
while (wait_for_heatup) {
  idle();
  wait_for_heatup = false;
  HOTEND_LOOP() {
    if (abs(thermalManager.degHotend(e) - thermalManager.degTargetHotend(e)) > 3) {
      wait_for_heatup = true;
      break;
    }
  }
}
```

The solution to the first issue is to ensure that the target temperature for the active extruder is set to a reasonable value when starting M600 in addition to verifying that the nozzle is actually at a reasonable value.

To correct the second issue, we just need to insert `if (!wait_for_heatup) break;` after the call to `idle()` in `resume_print`.

gruvin commented:
This kinda breaks stuff imo.
When "unpausing" using M108 GCODE, we now get the "PurgeMore / Continue" thing on the LCD. There's no GCODE to get past that. If the nozzle goes into standby, then the problem is once again compounded.

IMO, pause to change filament and pause just to pause, are two different things. They should not have been merged into one thing. Rather, M600 should call M125 -- or something of that nature. 

My point being that, as far as I can tell right now (bugfix-2.0.x #14729 ish. ) it seems that "simple" pause/resume with `NOZZLE_PARK_FEATURE` head parking from GCODE alone (`M125` or `M600` followed by `M108`?) is completely broken. I cannot even see a way to get M600 working without trying to unload filament or some variant of such.

there's so many other things that could make sense. Like, if the head remains parked for too long, don't just cool the head down but also unload filament but in such a way that an M108 break/continue reloads and self-primes -- without having to ask the user to confirm and choose and so on. (If that's already doable, I haven't been able to figure how, sorry.)

I looked into the code somewhat to see if I could contribute but sadly, it looks like my learning curve to make such an attempt would be weeks.

Thanks for your efforts. I know these comments can come across as a moaning session at times. Not my intention. :-)
